core/src/sigverify.rs

```diff
@@ -81,23 +85,55 @@ fn batch_size(batches: &[Packets]) -> usize {
     batches.iter().map(|p| p.packets.len()).sum()
 }
 
-pub fn get_packet_offsets(packet: &Packet, current_offset: u32) -> (u32, u32, u32, u32) {
-    let (sig_len, sig_size) = decode_len(&packet.data);
-    let msg_start_offset = sig_size + sig_len * size_of::<Signature>();
+// internal function to be unit-tested; should be used only by get_packet_offsets
+fn do_get_packet_offsets(packet: &Packet, current_offset: u32) -> (bool, u32, u32, u32, u32) {
+    // This directly reads the length of Transaction.signatures (serialized with short_vec)
+    let (sig_len_untrusted, sig_size) = decode_len(&packet.data);
+
+    // This directly reads MessageHeader.num_required_signatures (serialized with u8)
+    let msg_start_offset = sig_size + sig_len_untrusted * size_of::<Signature>();
+    // Using msg_start_offset which is based on sig_len_untrusted introduces uncertainity.
+    // Ultimately, the actual sigverify will determine the uncertainity.
+    let sig_len_maybe_trusted = packet.data[msg_start_offset] as usize;
 
-    let (_pubkey_len, pubkey_size) = decode_len(&packet.data[msg_start_offset..]);
+    let msg_header_size = serialized_size(&MessageHeader::default()).unwrap() as usize;
+
+    // This directly reads the length of Message.account_keys (serialized with short_vec)
+    let (_pubkey_len, pubkey_len_size) =
+        decode_len(&packet.data[(msg_start_offset + msg_header_size)..]);
 
     let sig_start = current_offset as usize + sig_size;
     let msg_start = current_offset as usize + msg_start_offset;
-    let pubkey_start =
-        msg_start + serialized_size(&MessageHeader::default()).unwrap() as usize + pubkey_size;
-
-    (
-        sig_len as u32,
-        sig_start as u32,
-        msg_start as u32,
-        pubkey_start as u32,
-    )
+    let pubkey_start = msg_start + msg_header_size + pubkey_len_size;
+
+    if sig_len_maybe_trusted == sig_len_untrusted {
+        (
+            true,
+            sig_len_maybe_trusted as u32,
+            sig_start as u32,
+            msg_start as u32,
+            pubkey_start as u32,
+        )
+    } else {
+        // a malformed packet is detected!!
+        (
+            false,
+            sig_len_untrusted as u32,
+            sig_start as u32,
+            msg_start as u32,
+            pubkey_start as u32,
+        )
+    }
+}
+
+pub fn get_packet_offsets(packet: &Packet, current_offset: u32) -> (u32, u32, u32, u32) {
```

Comment:
Can you create a return struct with named fields to replace this tuple of anonymous integers? (I know it predates your work)
Please update `do_get_packet_offsets` and `get_packet_offsets_from_tx` in tests to use it as well.

Reply:
Done with this patch series (021d597e4de641e7a21b9381998270763540ab8d, f10a36dfecfbb7d81a6ad2fcfc8c99f945371868, 2d0cb53b2190b66b84965e85c838929f76531793).

Because I'm new to Rust, any feedback well appreciated!